Source/JavaScriptCore/runtime/JSArray.cpp

```diff
@@ -1023,6 +1023,178 @@ JSString* JSArray::fastToString(JSGlobalObject* globalObject)
     RELEASE_AND_RETURN(scope, joiner.join(globalObject));
 }
 
+static uint64_t calculateFlattenedLength(JSArray* sourceArray, uint64_t sourceLength, uint64_t depth)
+{
+    uint64_t resultLength = 0;
+    IndexingType sourceType = sourceArray->indexingType();
+
+    switch (sourceType) {
+    case ArrayWithInt32:
+    case ArrayWithContiguous: {
+        auto* sourceBuffer = sourceArray->butterfly()->contiguous().data();
+        for (uint64_t i = 0; i < sourceLength; ++i) {
+            JSValue element = sourceBuffer[i].get();
+            if (!element) [[unlikely]]
+                continue;
+            if (sourceType == ArrayWithContiguous) {
+                if (depth > 0 && isJSArray(element)) {
+                    JSArray* elementArray = jsCast<JSArray*>(element);
+                    uint64_t newDepth = (depth == std::numeric_limits<uint64_t>::max()) ? depth : depth - 1;
+                    uint64_t flatLength = calculateFlattenedLength(elementArray, elementArray->length(), newDepth);
+                    if (flatLength == std::numeric_limits<uint64_t>::max()) [[unlikely]]
+                        return flatLength;
+                    resultLength += flatLength;
+                } else {
+                    if (element.isObject()) {
+                        auto elementObject = asObject(element);
+                        if (elementObject->isProxy()) [[unlikely]]
+                            return std::numeric_limits<uint64_t>::max();
+                    }
+                    resultLength++;
+                }
+            } else
+                resultLength++;
+        }
+        break;
+    }
+    case ArrayWithDouble: {
+        auto* sourceBuffer = sourceArray->butterfly()->contiguousDouble().data();
+        for (uint64_t i = 0; i < sourceLength; ++i) {
+            double value = sourceBuffer[i];
+            if (std::isnan(value)) [[unlikely]]
+                continue;
+            resultLength++;
+        }
+        break;
+    }
+    default:
+        return std::numeric_limits<uint64_t>::max();
+    }
+
+    return resultLength;
+}
+
+template<typename T>
+static uint64_t fastFlatIntoBuffer(JSGlobalObject* globalObject, T* resultBuffer, uint64_t& resultIndex, JSArray* sourceArray, uint64_t sourceLength, uint64_t depth, uint64_t vectorLength)
+{
+    VM& vm = globalObject->vm();
+    auto scope = DECLARE_CATCH_SCOPE(vm);
+
+    IndexingType sourceType = sourceArray->indexingType();
```

Comment:
We need to have recursion check via `vm.isSafeToRecurseSoft`. `flatIntoArray` is recursively calling, so it can be possible that we will have stackoverflow.